website/blog/2024-11-07-bedrock-chat-model-part-1/index.md

```diff
@@ -0,0 +1,1235 @@
+info:
+  title: Mars launch window API
+  version: 1.0.0
+  description: API to get the Mars launch window for Mission planning.
+paths:
+  /get-next-mars-launch-window:
+    get:
+      summary: Gets the next optimal launch window to Mars based on Hohmann transfer period.
+      description: Gets the next optimal launch window to Mars based on the rbital period difference in days for the Hohmann transfer to Mars.
+      operationId: getNextMarsLaunchWindow
+      responses:
+        "200":
+          description: Gets the next optimal launch window to Mars.
+          content:
+            "application/json":
+              schema:
+                type: object
+                properties:
+                  next_launch_window:
+                    type: string
+                    description: date of the next optimal launch window to Mars
+```
+
+### Action groups - Lamda function
+
+Here is the code deployment for action group's example Lambda:
+
+```python
+from datetime import datetime, timedelta
+import json
+
+
+def lambda_handler(event, context):
+    def _next_mars_launch_window():
+        current_date = datetime.now()
+        # Orbital period difference in days for the Hohmann transfer to Mars (~780 days)
+        hohmann_transfer_period = 780
+        # Last known optimal launch window to Mars (let's use a past known window for calculation)
+        # For example, let's use the successful Mars launch window on July 30, 2020
+        last_optimal_window = datetime(2020, 7, 30)
+        # Calculate the number of days since the last optimal window
+        days_since_last_window = (current_date - last_optimal_window).days
+        # Calculate the days until the next launch window
+        days_until_next_window = hohmann_transfer_period - (days_since_last_window % hohmann_transfer_period)
+        # Calculate the date of the next optimal launch window
+        next_window_date = current_date + timedelta(days=days_until_next_window)
+        return next_window_date.strftime("%Y-%m-%d")
+
+    response = {"next_launch_window": _next_mars_launch_window()}
+
+    response_body = {"application/json": {"body": json.dumps(response)}}
+
+    action_response = {
+        "actionGroup": event["actionGroup"],
+        "apiPath": event["apiPath"],
+        "httpMethod": event["httpMethod"],
+        "httpStatusCode": 200,
+        "responseBody": response_body,
+    }
+
+    session_attributes = event["sessionAttributes"]
+    prompt_session_attributes = event["promptSessionAttributes"]
+
+    return {
+        "messageVersion": "1.0",
+        "response": action_response,
+        "sessionAttributes": session_attributes,
+        "promptSessionAttributes": prompt_session_attributes,
+    }
+```
+
+Next, we are going to wrap Bedrock agent as a ChatModel so that we can register and load it for inference.
+
+## Writing ChatModel for Bedrock agent
+
+Here is the virtual env used for running the following example locally in **Python 3.12.7**:
```

Comment:
Do we need this? boto3 is notoriously stable :) Most of these dependencies are specified by MLflow or boto3. We could just specify a list of requirements to run it from the top-level packages